### PR TITLE
[WIP]add ability to add additional yum repos & add verification that sensitive files are deleted

### PR DESCRIPTION
**Pull request by M00nF1sh**
The changes in this PR is:
1.  add the ability to configure additional yum repo. sample usage:
     `make 1.14 additional_yum_repos='repo=guid-kernel-update,name=Kernel Updates,baseurl=http://amazonlinux.$awsregion.$awsdomain/2/core/2.0/$basearch/{guid},priority=10'` will add an additional repo for install kernel updates.
2.  add validation logic for AMI built, currently, the only validation include is
     * sensitive files are deleted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: scripts/install_additional_repos.sh [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    }
+  } 
+  Repo = "/etc/yum.repos.d/"vars["repo"]".repo"
+}
+{print "["vars["repo"]"]" > Repo}
+{print "name="vars["name"] > Repo}
+{print "baseurl="vars["baseurl"] > Repo }
+{if (length(vars["priority"]) != 0) print "priority="vars["priority"] > Repo }
+'
+echo "${ADDITIONAL_YUM_REPOS}" | awk "$AWK_CMD"
```
Comment (mogren) [inline scripts/install_additional_repos.sh]: Same here.

File: scripts/validate.sh [hunk 2]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
+#
+# do basic validations about AMI generated
```
Comment (mogren) [inline scripts/validate.sh]: Nit: `Do basic validation of the generated AMI`

File: scripts/validate.sh [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+#######################################
+# Validates file(or blob) don't exists
+# Arguments:
+#   an file name or blob
+# Returns:
+#   None
```
Comment (mogren) [inline scripts/validate.sh]:
```
# Validates that a file or blob doesn't exist
#
# Arguments:
#   a file name or blob
# Returns:
#   1 if a file exists, after printing an error
```

File: scripts/cleanup_additional_repos.sh [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    n = index($i, "="); 
+    if(n) { 
+      vars[substr($i, 1, n-1)] = substr($i, n + 1)
+    }
+  } 
+  Repo = "/etc/yum.repos.d/"vars["repo"]".repo"
+}
+{cmd="rm -f " Repo; system(cmd)}
+'
+echo "${ADDITIONAL_YUM_REPOS}" | awk "$AWK_CMD"
```
Comment (mogren) [inline scripts/cleanup_additional_repos.sh]: You could use a here-string...
```sh
awk "$AWK_CMD" <<< "${ADDITIONAL_YUM_REPOS}"
```

File: scripts/cleanup_additional_repos.sh [hunk 5]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env bash
+#
+# cleanup additional YUM repository(typically for security updates)
+# The format of ADDITIONAL_YUM_REPOS is like "repo=patches-repo,name=Install patches,baseurl=http://amazonlinux.$awsregion.$awsdomain/xxxx,priority=10"
+# Multiple yum repo can be specified separated by ';'
```
[the remark this answers is not in the document]
Comment (M00nF1sh) [reply scripts/cleanup_additional_repos.sh]: nice 🤣

File: scripts/cleanup_additional_repos.sh [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+if [ -z "${ADDITIONAL_YUM_REPOS}" ]; then
+  echo "no additional yum repo, skipping"
```
[the remark this answers is not in the document]
Comment (M00nF1sh) [reply scripts/cleanup_additional_repos.sh]: nice 🤣

File: scripts/install_additional_repos.sh [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+# Install additional YUM repository(typically for security updates)
+# The format of ADDITIONAL_YUM_REPOS is like "repo=patches-repo,name=Install patches,baseurl=http://amazonlinux.$awsregion.$awsdomain/xxxx,priority=10"
+# which will result in file '/etc/yum.repos.d/patches-repo.repo' with following content:
+# ```
+# [patches-repo]
+# name=Install patches
+# baseurl=http://amazonlinux.$awsregion.$awsdomain/xxxx
+# priority=10
+# ```
+# Multiple yum repo can be specified separated by ';'
```
[the remark this answers is not in the document]
Comment (M00nF1sh) [reply scripts/install_additional_repos.sh]: nice 🤣

File: scripts/install_additional_repos.sh [hunk 8]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
+#
+# Install additional YUM repository(typically for security updates)
+# The format of ADDITIONAL_YUM_REPOS is like "repo=patches-repo,name=Install patches,baseurl=http://amazonlinux.$awsregion.$awsdomain/xxxx,priority=10"
+# which will result in file '/etc/yum.repos.d/patches-repo.repo' with following content:
```
[the remark this answers is not in the document]
Comment (M00nF1sh) [reply scripts/install_additional_repos.sh]: thanks 🤣

File: scripts/install_additional_repos.sh [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    if(n) { 
+      vars[substr($i, 1, n-1)] = substr($i, n + 1)
+    }
+  } 
+  Repo = "/etc/yum.repos.d/"vars["repo"]".repo"
+}
+{print "["vars["repo"]"]" > Repo}
+{print "name="vars["name"] > Repo}
+{print "baseurl="vars["baseurl"] > Repo }
+{if (length(vars["priority"]) != 0) print "priority="vars["priority"] > Repo }
```
[the remark this answers is not in the document]
Comment (M00nF1sh) [reply scripts/install_additional_repos.sh]: fixed that space 👀

File: scripts/install_additional_repos.sh (hunk 1 above)
Comment (M00nF1sh) [reply scripts/install_additional_repos.sh]: fixed this space too 👀

File: scripts/validate.sh (hunk 2 above)
Comment (M00nF1sh) [reply scripts/validate.sh]: fixed 👀

File: scripts/validate.sh (hunk 3 above)
Comment (M00nF1sh) [reply scripts/validate.sh]: humm, it's not actually return but will exit the script with 1(error) :D , but fixed 👀

File: scripts/cleanup_additional_repos.sh (hunk 4 above)
Comment (M00nF1sh) [reply scripts/cleanup_additional_repos.sh]: changed :D i don't even know it can be used like this 🐰

File: scripts/validate.sh (hunk 3 above)
Comment (mogren) [reply scripts/validate.sh]: I meant that the whole validate.sh script exits with 1 😄

File: Makefile [hunk 10]
```diff
@@ -28,12 +28,12 @@ all: 1.11 1.12 1.13 1.14
...
 .PHONY: validate
 validate:
-	$(PACKER_BINARY) validate $(foreach packerVar,$(PACKER_VARIABLES), $(if $($(packerVar)),--var $(packerVar)=$($(packerVar)),)) eks-worker-al2.json
+	$(PACKER_BINARY) validate $(foreach packerVar,$(PACKER_VARIABLES), $(if $($(packerVar)),--var $(packerVar)='$($(packerVar))',)) eks-worker-al2.json
 
 .PHONY: k8s
 k8s: validate
 	@echo "$(T_GREEN)Building AMI for version $(T_YELLOW)$(kubernetes_version)$(T_GREEN) on $(T_YELLOW)$(arch)$(T_RESET)"
-	$(PACKER_BINARY) build $(foreach packerVar,$(PACKER_VARIABLES), $(if $($(packerVar)),--var $(packerVar)=$($(packerVar)),)) eks-worker-al2.json
+	$(PACKER_BINARY) build $(foreach packerVar,$(PACKER_VARIABLES), $(if $($(packerVar)),--var $(packerVar)='$($(packerVar))',)) eks-worker-al2.json
```
Comment (jqmichael) [inline Makefile]: Could you elaborate the motivation behind this change?

File: scripts/cleanup_additional_repos.sh [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    n = index($i, "="); 
+    if(n) { 
+      vars[substr($i, 1, n-1)] = substr($i, n + 1)
+    }
+  } 
+  Repo = "/etc/yum.repos.d/"vars["repo"]".repo"
+}
+{cmd="rm -f " Repo; system(cmd)}
+'
+sudo awk "$AWK_CMD" <<< "${ADDITIONAL_YUM_REPOS}"
```
Comment (jqmichael) [inline scripts/cleanup_additional_repos.sh]: This is really hard to read and maintain. Does Packer support high level programming language, i.e. Python?

File: Makefile (hunk 10 above)
Comment (M00nF1sh) [reply Makefile]: to allow the values contains space :D, like 'name=Kernel Updates'

File: scripts/cleanup_additional_repos.sh (hunk 11 above)
Comment (M00nF1sh) [reply scripts/cleanup_additional_repos.sh]: we need to install python runtime first.
TBH, it's similar to python programing, you'll have to write similar logic in python as well for parse such logic into a yum repo configuration.

Comment (jqmichael) [reply scripts/cleanup_additional_repos.sh]: To be more specific. 

Reader/maintainers needs to understand and remember awk `substr` is one based not the standard zero based.  https://www.gnu.org/software/gawk/manual/html_node/String-Functions.html#FOOT48
```
n = index($i, "="); 
    if(n) { 
      vars[substr($i, 1, n-1)] = substr($i, n + 1)
    }
```
Comparing with the Python version.
```
pair = split(field, '='):
vars[pair[0]] = dict[pair[1]]
```

Comment (M00nF1sh) [reply scripts/cleanup_additional_repos.sh]: to be fair..that's how shell sucks right....like 0 means true, 1 means false 🤣

**Pull request merged**